tools/svd2regs.py

```diff
@@ -0,0 +1,312 @@
+# usage: svd2regs.py [-h] (--mcu VENDOR MCU | --svd SVD) [--save FILE]
+#                    [--fmt] [--path PATH] [--args ARGS]
+#                    peripheral
+#
+# positional arguments:
+#   peripheral        Name of the Peripheral
+#
+# optional arguments:
+#   -h, --help        show this help message and exit
+#   --group, -g       Peripheral is a group with several instances
+#   --mcu VENDOR MCU  Vendor and MCU (Database from cmsis-svd)
+#   --svd SVD         Path to SVD-File
+#   --save FILE       Save generated Code to file
+#
+# rustfmt:
+# Format with rustfmt
+#
+#   --fmt ['ARG ..']  enable rustfmt with optional arguments
+#   --path PATH       path to rustfmt
+#
+# Examples:
+#   SIM peripheral from Database
+#     svd2regs.py SIM --mcu Freescale MK64F12
+#
+#   Format with rustfmt
+#     svd2regs.py SIM --svd mcu.svd --fmt
+#
+#   Format with rustfmt --force
+#     svd2regs.py SIM --svd mcu.svd --fmt '--force'
+#
+#   Format with rustfmt not in PATH
+#     svd2regs.py SIM --svd mcu.svd --fmt --path /home/tock/bin/
+#
+#   Save to file
+#     svd2regs.py SIM --svd mcu.svd --fmt '--force' --save src/peripherals.rs
+#
+#   With stdin pipe
+#     cat mcu.svd | svd2regs.py SIM --svd --fmt '--force' | tee src/mcu.rs
+#
+# Required Python Packages:
+#   cmsis-svd
+#   pydentifier
+#
+# Author: Stefan Hoelzl <stefan.hoelzl@posteo.de>
+
+import sys
+import argparse
+from subprocess import Popen, PIPE
+from xml.etree import ElementTree as ET
+
+from cmsis_svd.parser import SVDParser
+import pydentifier
+
+RUST_KEYWORDS = ["mod"]
+COMMENT_MAX_LENGTH = 80
+
+
+def comment(text):
+    return "/// {}".format(text[:COMMENT_MAX_LENGTH].strip())
+
+
+class CodeBlock(str):
+    TEMPLATE = ""
+
+    def __new__(cls, *args):
+        return cls.TEMPLATE.format(**cls.fields(*args))
+
+    @staticmethod
+    def fields(*args):
+        return {}
+
+
+class Includes(CodeBlock):
+    TEMPLATE = """
+use kernel::StaticRef;
+use kernel::common::regs::{{self, ReadOnly, ReadWrite, WriteOnly}};
```

Comment:
No need for double parentheses?

Reply:
it is a template string which gets formatted.
Places where values are inserted are marked with {}, so double parentheses do escape these to and get to single parentheses after inserting of the values.

When you look at other classes which derive from CodeBlock there are values to replace in the TEMPLATE.

Reply:
Ok, sorry!